test.c

```diff
@@ -1839,6 +1866,32 @@ response_status_cb (http_parser *p, const char *buf, size_t len)
   return 0;
 }
 
+int
+chunk_header_cb (http_parser *p)
+{
+  assert(p == parser);
+  if (p->content_length == 0) {
+    // Terminating chunk.  Don't increment num_chunks in this case.
```

Comment:
I added some tracing for processing the test.c:TWO_CHUNKS_MULT_ZERO_END message, here the output:

```
CALLBACK_NOTIFY_ message_begin
CALLBACK_DATA_ url mark[0]=/ len=25
CALLBACK_DATA_ header_field mark[0]=T len=17
CALLBACK_DATA_ header_value mark[0]=c len=7
CALLBACK_NOTIFY_ chunk_header
test.c:chunk_header_cb p->content_length=5
CALLBACK_DATA_ body mark[0]=h len=5
CALLBACK_NOTIFY_ chunk_complete
CALLBACK_NOTIFY_ chunk_header
test.c:chunk_header_cb p->content_length=6
CALLBACK_DATA_ body mark[0]=  len=6
CALLBACK_NOTIFY_ chunk_complete
CALLBACK_NOTIFY_ chunk_header
test.c:chunk_header_cb p->content_length=0
CALLBACK_NOTIFY_ chunk_complete
CALLBACK_NOTIFY_ message_complete
```

I hope this answers your question: the chunk_header_cb is fired even for the final len=0 (pseudo) chunk, and the tests validate the number of chunks excluding this final len=0 one. Here you could ask: do we really want chunk_header + chunk_complete events for this final length=0 chunk or not. I prefer having this callback (since the parser's job is to markup byte blocks in the byte stream with their meaning, including the meaning for 0 or 000 as a chunk list terminator).